File: android/app/src/main/java/com/whoami/today/app/widget/WidgetData.java
package com.whoami.today.app.widget;

import org.json.JSONArray;
import org.json.JSONObject;
import java.util.ArrayList;
import java.util.List;

public class WidgetData {
    public MyCheckIn myCheckIn;
    public List<FriendUpdate> friendsWithUpdates;
    public List<PlaylistSong> sharedPlaylists;
    public QuestionOfDay questionOfDay;

    public static WidgetData fromJson(String jsonString) {
        try {
            JSONObject json = new JSONObject(jsonString);
            WidgetData data = new WidgetData();

            // Parse my check-in
            if (json.has("my_check_in") && !json.isNull("my_check_in")) {
                data.myCheckIn = MyCheckIn.fromJson(json.getJSONObject("my_check_in"));
            }

            // Parse friends with updates
            data.friendsWithUpdates = new ArrayList<>();
            if (json.has("friends_with_updates")) {
                JSONArray friendsArray = json.getJSONArray("friends_with_updates");
                for (int i = 0; i < friendsArray.length(); i++) {
                    data.friendsWithUpdates.add(FriendUpdate.fromJson(friendsArray.getJSONObject(i)));
                }
            }

            // Parse shared playlists
            data.sharedPlaylists = new ArrayList<>();
            if (json.has("shared_playlists")) {
                JSONArray playlistsArray = json.getJSONArray("shared_playlists");
                for (int i = 0; i < playlistsArray.length(); i++) {
                    data.sharedPlaylists.add(PlaylistSong.fromJson(playlistsArray.getJSONObject(i)));
                }
            }

            // Parse question of day
            if (json.has("question_of_day") && !json.isNull("question_of_day")) {
                data.questionOfDay = QuestionOfDay.fromJson(json.getJSONObject("question_of_day"));
            }

            return data;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    public static class MyCheckIn {
        public int id;
        public String mood;
        public String socialBattery;
        public String trackId;
        public String albumImageUrl;

        public static MyCheckIn fromJson(JSONObject json) {
            try {
                MyCheckIn checkIn = new MyCheckIn();
                checkIn.id = json.optInt("id");
                checkIn.mood = json.optString("mood", "🤔");
                checkIn.socialBattery = json.optString("social_battery", "half");
                checkIn.trackId = json.optString("track_id", "");
                checkIn.albumImageUrl = json.optString("album_image_url", null);
                return checkIn;
            } catch (Exception e) {
                e.printStackTrace();
                return null;
            }
        }

        /** Matches SocialBatteryChipAssets in TS: completely_drained, low, needs_recharge, moderately_social, fully_charged, super_social */
        public String getBatteryEmoji() {
            if (socialBattery == null) return "🪫";
            switch (socialBattery) {
                case "completely_drained": return "💤";
                case "low": return "🪫";
                case "needs_recharge": return "🔌";
                case "moderately_social": return "🔋";
                case "fully_charged": return "🚀";
                case "super_social": return "🤩";
                default: return "🔋";
            }
        }
    }

    public static class FriendUpdate {
        public int id;
        public String username;
        public String profilePic;
        public String profileImage;

        public static FriendUpdate fromJson(JSONObject json) {
            try {
                FriendUpdate friend = new FriendUpdate();
                friend.id = json.optInt("id");
                friend.username = json.optString("username", "");
                friend.profilePic = json.optString("profile_pic", "#AABBCC");
                friend.profileImage = json.optString("profile_image", null);
                // Support nested "user" object (e.g. API returns { "user": { "username": "x" } })
                if (json.has("user") && !json.isNull("user")) {
                    JSONObject user = json.getJSONObject("user");
                    if (friend.username.isEmpty()) friend.username = user.optString("username", "");
                    if (friend.profileImage == null || friend.profileImage.isEmpty()) friend.profileImage = user.optString("profile_image", null);
                    if ("#AABBCC".equals(friend.profilePic)) friend.profilePic = user.optString("profile_pic", friend.profilePic);
                }
                return friend;
            } catch (Exception e) {
                e.printStackTrace();
                return null;
            }
        }
    }

    public static class PlaylistSong {
        public int id;
        public PlaylistUser user;
        public String trackId;
        public String createdAt;

        public static PlaylistSong fromJson(JSONObject json) {
            try {
                PlaylistSong song = new PlaylistSong();
                song.id = json.optInt("id");
                song.user = PlaylistUser.fromJson(json.getJSONObject("user"));
                song.trackId = json.optString("track_id", "");
                song.createdAt = json.optString("created_at", "");
                return song;
            } catch (Exception e) {
                e.printStackTrace();
                return null;
            }
        }
    }

    public static class PlaylistUser {
        public int id;
        public String username;
        public String profilePic;
        public String profileImage;

        public static PlaylistUser fromJson(JSONObject json) {
            try {
                PlaylistUser user = new PlaylistUser();
                user.id = json.optInt("id");
                user.username = json.optString("username", "");
                user.profilePic = json.optString("profile_pic", "#AABBCC");
                user.profileImage = json.optString("profile_image", null);
                return user;
            } catch (Exception e) {
                e.printStackTrace();
                return null;
            }
        }
    }

    public static class QuestionOfDay {
        public String id;
        public String question;
        public String content;
        public String createdAt;

        public static QuestionOfDay fromJson(JSONObject json) {
            try {
                QuestionOfDay q = new QuestionOfDay();
                q.id = json.optString("id", "");
                q.question = json.optString("question", "");
                q.content = json.optString("content", "");
                q.createdAt = json.optString("created_at", "");
                return q;
            } catch (Exception e) {
                e.printStackTrace();
                return null;
            }
        }

        public String getDeepLink() {
            return "whoami://app/questions/" + id + "/new";
        }
    }
}

File: ios/WhoAmITodayWidgetExtension_DEPRECATED/Models/WidgetData.swift
import Foundation

// MARK: - Widget Display Data
struct WidgetData: Codable {
    let myCheckIn: MyCheckIn?
    let friendsWithUpdates: [FriendUpdate]
    let sharedPlaylists: [PlaylistSong]
    let questionOfDay: QuestionOfDay?
    let lastUpdated: Date

    static var placeholder: WidgetData {
        WidgetData(
            myCheckIn: MyCheckIn(
                id: 1,
                isActive: true,
                createdAt: "",
                mood: "🤔",
                socialBattery: nil,  // No battery check-in → widget shows + for My Battery
                description: "",
                trackId: "",
                albumImageUrl: nil
            ),
            friendsWithUpdates: [
                FriendUpdate(id: 1, username: "Matter123", profilePic: "#AABBCC", profileImage: nil),
                FriendUpdate(id: 2, username: "Kipler2323", profilePic: "#FFAABB", profileImage: nil)
            ],
            sharedPlaylists: [
                PlaylistSong(id: 1, user: PlaylistUser(id: 1, username: "User1", profilePic: "#AABBCC", profileImage: nil), trackId: "track1", createdAt: ""),
                PlaylistSong(id: 2, user: PlaylistUser(id: 2, username: "User2", profilePic: "#BBCCDD", profileImage: nil), trackId: "track2", createdAt: ""),
                PlaylistSong(id: 3, user: PlaylistUser(id: 3, username: "User3", profilePic: "#CCDDEE", profileImage: nil), trackId: "track3", createdAt: ""),
                PlaylistSong(id: 4, user: PlaylistUser(id: 4, username: "User4", profilePic: "#DDEEFF", profileImage: nil), trackId: "track4", createdAt: ""),
                PlaylistSong(id: 5, user: PlaylistUser(id: 5, username: "User5", profilePic: "#EEFFAA", profileImage: nil), trackId: "track5", createdAt: "")
            ],
            questionOfDay: QuestionOfDay(
                id: "1",
                question: "What was a funny thing that happened today?",
                content: "What was a funny thing that happened today?",
                createdAt: nil
            ),
            lastUpdated: Date()
        )
    }

    static var empty: WidgetData {
        WidgetData(
            myCheckIn: nil,
            friendsWithUpdates: [],
            sharedPlaylists: [],
            questionOfDay: nil,
            lastUpdated: Date()
        )
    }
}

// MARK: - My Check-In (from /api/user/me/profile -> check_in field)
struct MyCheckIn: Codable {
    let id: Int
    let isActive: Bool
    let createdAt: String
    let mood: String              // "I feel" emoji/text
    let socialBattery: String?    // "My Battery" - "fully_charged", "half", "low", etc.
    let description: String
    let trackId: String           // "My Music" - Spotify track ID
    let albumImageUrl: String?    // Album cover image URL from Spotify

    enum CodingKeys: String, CodingKey {
        case id
        case isActive = "is_active"
        case createdAt = "created_at"
        case mood
        case socialBattery = "social_battery"
        case description
        case trackId = "track_id"
        case albumImageUrl = "album_image_url"
    }

    // Helper to get display emoji for "I feel"
    var feelingDisplay: String {
        mood.isEmpty ? "🤔" : mood
    }

    // Helper to get battery emoji based on level
    var batteryDisplay: String {
        guard let battery = socialBattery else { return "🪫" }
        switch battery {
        case "fully_charged": return "🔋"
        case "half", "medium": return "🔋"
        case "low": return "🪫"
        case "empty": return "🪫"
        default: return "🔋"
        }
    }

    // Helper for music display
    var musicDisplay: String {
        trackId.isEmpty ? "🎵" : "🎵"
    }
}

// MARK: - API Response: Friend with Updates (from /api/user/friends/updates/)
struct FriendUpdate: Codable, Identifiable {
    let id: Int
    let username: String
    let profilePic: String      // Hex color code (e.g. "#AABBCC")
    let profileImage: String?   // URL to image or null

    enum CodingKeys: String, CodingKey {
        case id
        case username
        case profilePic = "profile_pic"
        case profileImage = "profile_image"
    }
}

// MARK: - API Response: Playlist Song (from /api/playlist/feed)
struct PlaylistSong: Codable, Identifiable {
    let id: Int
    let user: PlaylistUser
    let trackId: String
    let createdAt: String

    enum CodingKeys: String, CodingKey {
        case id
        case user
        case trackId = "track_id"
        case createdAt = "created_at"
    }
}

struct PlaylistUser: Codable {
    let id: Int
    let username: String
    let profilePic: String
    let profileImage: String?

    enum CodingKeys: String, CodingKey {
        case id
        case username
        case profilePic = "profile_pic"
        case profileImage = "profile_image"
    }
}

// MARK: - API Response: My Profile (from /api/user/me/)
struct MyProfileResponse: Codable {
    let id: Int
    let username: String
    let checkIn: MyCheckIn?

    enum CodingKeys: String, CodingKey {
        case id
        case username
        case checkIn = "check_in"
    }
}

// MARK: - Question of the Day
struct QuestionOfDay: Codable {
    let id: String
    let question: String
    let content: String
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case question
        case content
        case createdAt = "created_at"
    }

    var deepLink: String {
        "whoami://app/questions/\(id)/new"
    }
}

// MARK: - Paginated Response wrapper
struct PaginatedResponse<T: Codable>: Codable {
    let count: Int
    let next: String?
    let previous: String?
    let results: [T]
}

// MARK: - General Questions Response
struct QuestionGroup: Codable {
    let date: String
    let questions: [GeneralQuestion]
}

struct GeneralQuestion: Codable {
    let type: String
    let id: Int
    let content: String
    let createdAt: String
    let isAdminQuestion: Bool
    let selectedDates: [String]
    let selected: Bool

    enum CodingKeys: String, CodingKey {
        case type
        case id
        case content
        case createdAt = "created_at"
        case isAdminQuestion = "is_admin_question"
        case selectedDates = "selected_dates"
        case selected
    }

    // Convert GeneralQuestion to QuestionOfDay
    func toQuestionOfDay() -> QuestionOfDay {
        return QuestionOfDay(
            id: String(id),
            question: content,
            content: content,
            createdAt: createdAt
        )
    }
}
